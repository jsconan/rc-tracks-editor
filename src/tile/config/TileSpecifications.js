/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gives the specifications for the track tiles with respect to the initial constraints.
 */
export class TileSpecifications {
    /**
     * Defines the specifications for the track tiles from the given constraints.
     * @param {object} config - A set of config options.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @param {boolean} unlockRatio - Unlock the maximum size ratio some tiles are applying.
     */

    constructor({ laneWidth = 20, barrierWidth = 1, barrierChunks = 4, maxRatio = 4, unlockRatio = false } = {}) {
        this.setLaneWidth(laneWidth);
        this.setBarrierWidth(barrierWidth);
        this.setBarrierChunks(barrierChunks);
        this.setMaxRatio(maxRatio);
        this.setUnlockRatio(unlockRatio);
    }

    /**
     * The context identifier to retrieve the config from the context.
     * @type {string}
     */
    get contextId() {
        return this.constructor.CONTEXT_ID;
    }

    /**
     * The length of a tile with respect to the initial constraints.
     * @type {number}
     */
    get length() {
        return this.laneWidth * 1.25 + this.barrierWidth * 2;
    }

    /**
     * The width of a tile with respect to the initial constraints.
     * @type {number}
     */
    get width() {
        return this.laneWidth + this.barrierWidth * 2;
    }

    /**
     * The length of a barrier chunk with respect to the initial constraints.
     * @type {number}
     */
    get barrierLength() {
        return this.length / this.barrierChunks;
    }

    /**
     * The padding around the width of the tile to align with its length.
     * Note: tiles must fit in a square so that they can be flipped and rotated indifferently.
     * @type {number}
     */
    get padding() {
        return this.laneWidth / 8;
    }

    /**
     * Sets the width of the track lane (the distance between the barriers).
     * @param {number} laneWidth - The width of the track lane.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setLaneWidth(laneWidth) {
        this.laneWidth = Math.abs(laneWidth);

        return this;
    }

    /**
     * Sets the width of the barriers.
     * @param {number} barrierWidth - The width of the barriers.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierWidth(barrierWidth) {
        this.barrierWidth = Math.abs(barrierWidth);

        return this;
    }

    /**
     * Sets the number of barrier chunks per tile.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierChunks(barrierChunks) {
        this.barrierChunks = Math.abs(Math.round(barrierChunks) || 1);

        return this;
    }

    /**
     * Sets the maximum value for size ratios.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setMaxRatio(maxRatio) {
        this.maxRatio = Math.abs(Math.round(maxRatio) || 1);

        return this;
    }

    /**
     * Locks or unlocks the maximum size ratio some tiles are applying.
     * @param {boolean} unlockRatio - Whether or not unlock the size ratio.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setUnlockRatio(unlockRatio) {
        this.unlockRatio = !!unlockRatio;

        return this;
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The specifications object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * The context identifier to retrieve the config from the context.
 * @constant {string} TileSpecifications.CONTEXT_ID
 */
Object.defineProperty(TileSpecifications, 'CONTEXT_ID', {
    value: 'tileSpecifications',
    writable: false,
    enumerable: true,
    configurable: true
});
