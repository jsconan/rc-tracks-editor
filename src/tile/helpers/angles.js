/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Vector2D } from '../../core/models';

const lastQuadrant = Vector2D.STRAIGHT_ANGLE + Vector2D.RIGHT_ANGLE;
const shiftAngle = Vector2D.RIGHT_ANGLE / 2;

/**
 * Returns the angle of the closest quadrant with respect to an angle.
 * @param {number} angle - The angle from which get the quadrant angle.
 * @returns {number} - Returns the angle of the closest quadrant.
 */
export const quadrantAngle = angle => {
    const edge = Vector2D.quadrant(angle + shiftAngle) * Vector2D.RIGHT_ANGLE;
    if (!edge && angle > lastQuadrant) {
        return Vector2D.CIRCLE;
    }
    return edge;
};

/**
 * Returns the angle of the closest quadrant included in a range of angles.
 * The range is not expected to be higher than 90 degrees wide.
 * @param {number} start - The start angle of the range.
 * @param {number} end - The end angle of the range.
 * @returns {number} - The quadrant angle contained by the range. It may be null.
 */
export const quadrantRange = (start, end) => {
    const startQuadrant = quadrantAngle(start);
    const endQuadrant = quadrantAngle(end);

    if (start < startQuadrant && startQuadrant < end) {
        return startQuadrant;
    }

    if (start < endQuadrant && endQuadrant < end) {
        return endQuadrant;
    }

    return null;
};
