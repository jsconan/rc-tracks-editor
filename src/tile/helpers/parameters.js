/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_ENLARGED_TYPE, CURVED_TILE_TYPE, STRAIGHT_TILE_TYPE, validateType } from './tiles';

/**
 * Computes the parameters for rendering a tile at the expected position.
 * @type {object}
 * @private
 */
const tileParameters = {
    /**
     * Computes the parameters for rendering a curved tile at the expected position.
     * @param {CurvedTileModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [CURVED_TILE_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const width = specs.width;
        const barrierWidth = specs.barrierWidth;
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius() - barrierWidth;
        const innerChunks = model.getInnerBarrierChunks();
        const outerChunks = model.getOuterBarrierChunks();
        const curveAngle = model.getCurveAngle();
        const curveCenter = model.getCurveCenter(tileX, tileY);

        const innerX = curveCenter.x + innerRadius;
        const innerY = curveCenter.y;
        const outerX = curveCenter.x + outerRadius;
        const outerY = curveCenter.y;

        const ground = {
            cx: curveCenter.x,
            cy: curveCenter.y,
            width,
            radius: innerRadius,
            angle: curveAngle,
            start: 0
        };
        const innerBarrier = {
            chunks: innerChunks,
            width: barrierWidth,
            radius: innerRadius,
            angle: curveAngle,
            left: innerX,
            top: innerY,
            shift: 0
        };
        const outerBarrier = {
            chunks: outerChunks,
            width: barrierWidth,
            radius: outerRadius,
            angle: curveAngle,
            left: outerX,
            top: outerY,
            shift: 1
        };

        return {
            ground,
            innerBarrier,
            outerBarrier
        };
    },

    /**
     * Computes the parameters for rendering a curved tile enlarged at the expected position.
     * @param {CurvedTileEnlargedModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [CURVED_TILE_ENLARGED_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const width = specs.width;
        const barrierLength = specs.barrierLength;
        const barrierWidth = specs.barrierWidth;
        const side = model.getCurveSide();
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius() - barrierWidth;
        const sideChunks = model.getSideBarrierChunks();
        const innerChunks = model.getInnerBarrierChunks();
        const outerChunks = model.getOuterBarrierChunks();
        const curveAngle = model.getCurveAngle();
        const curveCenter = model.getCurveCenter(tileX, tileY);
        const outerBarrierPosition = width - barrierWidth;

        const { x: innerX, y: innerY } = curveCenter.addScalarX(innerRadius);
        const { x: outerX, y: outerY } = curveCenter.addCoord(
            side + outerRadius,
            innerRadius + outerBarrierPosition - outerRadius
        );
        const { x: horizontalX, y: horizontalY } = curveCenter.addScalarY(innerRadius + outerBarrierPosition);
        const { x: verticalX, y: verticalY } = curveCenter.addScalarX(innerRadius + outerBarrierPosition);

        const ground = {
            cx: curveCenter.x,
            cy: curveCenter.y,
            width,
            side,
            radius: innerRadius
        };
        const innerBarrier = {
            chunks: innerChunks,
            width: barrierWidth,
            radius: innerRadius,
            angle: curveAngle,
            left: innerX,
            top: innerY,
            shift: 0
        };
        const outerBarrier = {
            chunks: outerChunks,
            width: barrierWidth,
            radius: outerRadius,
            angle: curveAngle,
            left: outerX,
            top: outerY,
            shift: 1
        };
        const horizontalBarrier = {
            chunks: sideChunks,
            width: barrierWidth,
            length: barrierLength,
            left: horizontalX,
            top: horizontalY,
            shift: 0,
            vertical: false
        };
        const verticalBarrier = {
            chunks: sideChunks,
            width: barrierWidth,
            length: barrierLength,
            left: verticalX,
            top: verticalY,
            shift: 1,
            vertical: true
        };

        return {
            ground,
            innerBarrier,
            outerBarrier,
            horizontalBarrier,
            verticalBarrier
        };
    },

    /**
     * Computes the parameters for rendering a straight tile at the expected position.
     * @param {StraightTileModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [STRAIGHT_TILE_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const barrierLength = specs.barrierLength;
        const barrierWidth = specs.barrierWidth;
        const width = model.width;
        const height = model.length;
        const chunks = model.getSideBarrierChunks();
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius();
        const curveCenter = model.getCurveCenter(tileX, tileY);

        const leftX = curveCenter.x + innerRadius;
        const leftY = curveCenter.y;
        const rightX = curveCenter.x + outerRadius - barrierWidth;
        const rightY = curveCenter.y;
        const vertical = true;

        const ground = {
            x: leftX,
            y: leftY,
            width,
            height
        };
        const leftBarrier = {
            chunks,
            width: barrierWidth,
            length: barrierLength,
            left: leftX,
            top: leftY,
            shift: 0,
            vertical
        };
        const rightBarrier = {
            chunks,
            width: barrierWidth,
            length: barrierLength,
            left: rightX,
            top: rightY,
            shift: 1,
            vertical
        };

        return { ground, leftBarrier, rightBarrier };
    }
};

/**
 * Computes the parameters for rendering a tile at the expected position.
 * @param {TileModel} model - A reference to the tile model.
 * @param {number} tileX - The X-coordinate of the tile.
 * @param {number} tileY - The Y-coordinate of the tile.
 * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
 * @throws {TypeError} - If the given type is not valid.
 */

export function getTileParameters(model, tileX, tileY) {
    const { type } = model;
    validateType(type);
    return tileParameters[type](model, tileX, tileY);
}

/**
 * @typedef {import('../models').TileModel} TileModel
 */

/**
 * @typedef {import('../models').CurvedTileModel} CurvedTileModel
 */

/**
 * @typedef {import('../models').CurvedTileEnlargedModel} CurvedTileEnlargedModel
 */

/**
 * @typedef {import('../models').StraightTileModel} StraightTileModel
 */
