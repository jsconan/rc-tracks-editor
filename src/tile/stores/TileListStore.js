/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { EventStore } from '../../core/stores';
import { TileList } from '../models';

/**
  Implements a store bound to a list of tiles.
  The store will be updated each time the list is modified.
 */
export class TileListStore extends EventStore {
    /**
     * Creates a store bound to a list of tiles. The store will be updated each time the list is modified.
     * @param {TileList} [boundTo] - The list of tiles to bind with the store.
     * @param {eventStoreUpdater} update - A callback that will be called for setting the store each time a listened event is emitted.
     * If it is omitted, the bound list of tiles will be set on each update.
     * @throws {TypeError} - If the given object is not a TileList.
     */
    constructor(boundTo = null, update = null) {
        super(
            [
                // List events
                'set',
                'add',
                'delete',
                'clear',
                'load',
                // TileList events
                'specs',
                'update'
            ],
            boundTo,
            update
        );
    }

    /**
     * Validates that the given object is a TileList and can be bound to store.
     * @param {TileList} list - The list of tiles to bind with the store.
     * @returns {TileListStore} - Chains the instance.
     * @throws {TypeError} - If the given object is not a TileList.
     */
    validate(list) {
        TileList.validateInstance(list);
        return super.validate(list);
    }
}

/**
 * @typedef {import('../../core/stores').eventStoreUpdater} eventStoreUpdater
 */
