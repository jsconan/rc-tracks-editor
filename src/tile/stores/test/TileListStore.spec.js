/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_TYPE } from '../../helpers';
import { TileSpecifications } from '../../config';
import { TileList } from '../../models';
import { TileListStore } from '../TileListStore.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications({ laneWidth, barrierWidth, barrierChunks });

describe('tileListStore', () => {
    it('is a class', () => {
        expect(TileListStore).toEqual(expect.any(Function));
    });

    describe('throws error', () => {
        it('if the given object is not a TileList', () => {
            expect(() => new TileListStore({})).toThrow('The object must be an instance of TileList!');

            const store = new TileListStore();
            expect(() => store.validate({})).toThrow('The object must be an instance of TileList!');
            expect(() => store.bind({})).toThrow('The object must be an instance of TileList!');
        });

        it('if the given object is not a TileListStore', () => {
            expect(() => TileListStore.validateInstance({})).toThrow('The object must be an instance of TileListStore');
            expect(() => TileListStore.validateInstance(new TileListStore())).not.toThrow();
        });
    });

    it('creates a store', () => {
        const list = new TileList(specs);
        const store = new TileListStore(list);

        expect(store).toEqual(expect.any(Object));
        expect(store.subscribe).toEqual(expect.any(Function));
        expect(store.notify).toEqual(expect.any(Function));
        expect(store.bind).toEqual(expect.any(Function));
        expect(store.unbind).toEqual(expect.any(Function));
        expect(store.boundTo).toBe(list);
    });

    it('updates the store each time the list of tiles is modified', () => {
        const list = new TileList(specs);
        const store = new TileListStore(list);

        expect(store.boundTo).toBe(list);

        const callback = jest.fn().mockImplementation(value => {
            expect(value).toBe(list);
        });

        store.subscribe(callback);

        const tile = list.append();
        list.remove(tile.id);
        list.update();

        expect(callback).toHaveBeenCalledTimes(4);
    });

    it('updates the store when calling the notify API', () => {
        const list = new TileList(specs);
        let currentEvent = null;
        const updateCallback = jest.fn().mockImplementation((value, event) => {
            expect(event).toBe(currentEvent);
            expect(value).toBe(list);
            return value;
        });
        const store = new TileListStore(list, updateCallback);

        expect(store.boundTo).toBe(list);

        const subscriberCallback = jest.fn().mockImplementation(value => {
            expect(value).toBe(list);
        });

        store.subscribe(subscriberCallback);

        currentEvent = 'test';
        store.notify(currentEvent);

        expect(updateCallback).toHaveBeenCalledTimes(2);
        expect(subscriberCallback).toHaveBeenCalledTimes(2);
    });

    it('can bind the list of tiles later', () => {
        const list = new TileList(specs);
        const store = new TileListStore();

        const callback = jest.fn().mockImplementation(value => {
            if (value) {
                expect(value).toBe(list);
            }
        });

        store.subscribe(callback);

        list.update();

        expect(store.boundTo).toBeNull();
        expect(store.bind(list)).toBe(store);
        expect(store.boundTo).toBe(list);

        list.update();

        expect(callback).toHaveBeenCalledTimes(3);
    });

    it('can replace the event emitter later', () => {
        const list1 = new TileList(specs);
        const list2 = new TileList(specs);
        const store = new TileListStore(list1);

        const callback1 = jest.fn().mockImplementation(value => {
            expect(value).toBe(list1);
        });

        const callback2 = jest.fn().mockImplementation(value => {
            expect(value).toBe(list2);
        });

        const unsubscribe = store.subscribe(callback1);

        unsubscribe();

        expect(store.boundTo).toBe(list1);

        store.bind(list2);
        store.subscribe(callback2);

        expect(store.boundTo).toBe(list2);

        list2.update();

        expect(callback1).toHaveBeenCalledTimes(1);
        expect(callback2).toHaveBeenCalledTimes(2);
    });

    it('accepts a callback to set the store each time a listed event is emitted', () => {
        const list = new TileList(specs);
        const data = {};

        let currentEvent = null;
        const update = jest.fn().mockImplementation((value, event) => {
            expect(event).toBe(currentEvent);
            expect(value).toBe(list);
            return data;
        });

        const store = new TileListStore(list, update);

        const callback = jest.fn().mockImplementation(value => {
            expect(value).toBe(data);
        });

        store.subscribe(callback);

        currentEvent = 'update';
        list.update();

        expect(update).toHaveBeenCalledTimes(2);
        expect(callback).toHaveBeenCalledTimes(2);
    });

    it('can release the listeners', () => {
        const list = new TileList(specs);
        const store = new TileListStore(list);

        const callback = jest.fn().mockImplementation(value => {
            expect(value).toBe(list);
        });

        store.subscribe(callback);

        list.update();

        expect(store.boundTo).toBe(list);

        expect(store.unbind()).toBe(store);

        expect(store.boundTo).toBeNull();

        list.update();

        expect(callback).toHaveBeenCalledTimes(2);
    });

    it('does not impact the other listeners when releasing', () => {
        const list = new TileList(specs);
        const store = new TileListStore(list);

        const listener = jest.fn();

        list.on('update', listener);

        list.update();

        store.unbind();

        list.update();

        expect(listener).toHaveBeenCalledTimes(2);
    });

    it('can notify changes applied to the list', () => {
        const list = new TileList(specs);
        const store = new TileListStore(list);

        const callback = jest.fn().mockImplementation(value => {
            expect(value).toBe(list);
            expect(value).toMatchSnapshot();
        });

        const unsubscribe = store.subscribe(callback); // callback called

        list.setSpecs(specs);
        const tile1 = list.append(CURVED_TILE_TYPE); // callback called
        const tile2 = list.prepend(CURVED_TILE_TYPE); // callback called
        list.remove(tile1.id); // callback called
        const tile3 = list.replace(tile2.id); // callback called
        list.insertBefore(tile3.id); // callback called
        list.insertAfter(tile3.id); // callback called
        const data = list.export();
        list.clear(); // callback called
        list.import(data); // callback called
        list.setSpecs(new TileSpecifications({ laneWidth: 10, barrierWidth: 1, barrierChunks: 2 })); // callback called

        list.update(); // callback called

        unsubscribe();
        list.append();

        expect(callback).toHaveBeenCalledTimes(11);
    });
});
