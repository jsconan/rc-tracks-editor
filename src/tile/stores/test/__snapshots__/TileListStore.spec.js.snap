// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tileListStore can notify changes applied to the list 1`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 2`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    CurvedTileModel {
      "direction": "right",
      "id": "id-1",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 3`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    CurvedTileModel {
      "direction": "right",
      "id": "id-2",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    CurvedTileModel {
      "direction": "right",
      "id": "id-1",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 4`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    CurvedTileModel {
      "direction": "right",
      "id": "id-2",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 5`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-3",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 6`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-4",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-3",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 7`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-4",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-3",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-5",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 8`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 9`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-6",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-7",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-8",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 10`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-6",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-7",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-8",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 2,
    "barrierWidth": 1,
    "laneWidth": 10,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;

exports[`tileListStore can notify changes applied to the list 11`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": [
    StraightTileModel {
      "direction": "right",
      "id": "id-6",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-7",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-8",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 2,
        "barrierWidth": 1,
        "laneWidth": 10,
        "maxRatio": 4,
        "unlockRatio": false,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 2,
    "barrierWidth": 1,
    "laneWidth": 10,
    "maxRatio": 4,
    "unlockRatio": false,
  },
}
`;
